lib/prom/prom.go

```diff
@@ -0,0 +1,121 @@
+package prom
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	vegeta "github.com/tsenart/vegeta/v12/lib"
+)
+
+//PrometheusMetrics vegeta metrics observer with exposition as Prometheus metrics endpoint
+type PrometheusMetrics struct {
+	requestSecondsHistogram *prometheus.HistogramVec
+	requestBytesInCounter   *prometheus.CounterVec
+	requestBytesOutCounter  *prometheus.CounterVec
+	requestFailCounter      *prometheus.CounterVec
+	listenPort              net.Listener
+}
+
+//NewPrometheusMetrics same as NewPrometheusMetricsWithParams with default params:
+//bindHost=0.0.0.0, bindPort=8880 and metricsPath=/metrics
+func NewPrometheusMetrics() (PrometheusMetrics, error) {
+	return NewPrometheusMetricsWithParams("0.0.0.0", 8880, "/metrics")
+}
+
+// NewPrometheusMetricsWithParams start a new Prometheus observer instance for exposing
+// metrics to Prometheus Servers.
+// Only one PrometheusMetrics can be instantiated at a time because of the underlaying
+// mechanisms of promauto.
+// Some metrics are requests/s, bytes in/out/s and failures/s
+// Options are:
+//   - bindHost: host to bind the listening socket to
+//   - bindPort: port to bind the listening socket to
+//   - metricsPath: http path that will be used to get metrics
+// For example, after using NewPrometheusMetricsWithParams("0.0.0.0", 8880, "/metrics"),
+// during an "attack" you can call "curl http://127.0.0.0:8880/metrics" to see current metrics.
+// This endpoint can be configured in scrapper section of your Prometheus server.
+func NewPrometheusMetricsWithParams(bindHost string, bindPort int, metricsPath string) (PrometheusMetrics, error) {
+
+	pm := PrometheusMetrics{}
+
+	pm.requestSecondsHistogram = promauto.NewHistogramVec(prometheus.HistogramOpts{
```

Comment:
Using promauto means we write to the global package level Prometheus registry. This may have unexpected effects when vegeta is used as a library. Is there a way to be explicit about adding the metrics to a specific registry?

Reply:
Sure.

Reply:
done